baseapp/baseapp.go

```diff
@@ -703,37 +704,35 @@ func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, mode runTxMode) (*s
 		}
 
 		var (
-			msgEvents sdk.Events
-			msgResult *sdk.Result
-			msgFqName string
-			err       error
+			msgResult      *sdk.Result
+			msgEventAction string // name to use as value in event `message.action`
+			err            error
 		)
 
-		if svcMsg, ok := msg.(sdk.ServiceMsg); ok {
-			msgFqName = svcMsg.MethodName
-			handler := app.msgServiceRouter.Handler(msgFqName)
-			if handler == nil {
-				return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "unrecognized message service method: %s; message index: %d", msgFqName, i)
-			}
-			msgResult, err = handler(ctx, svcMsg.Request)
-		} else {
+		if handler := app.msgServiceRouter.Handler(msg); handler != nil {
+			// ADR 031 request type routing
+			msgResult, err = handler(ctx, msg)
+			msgEventAction = sdk.MsgTypeURL(msg)
+		} else if legacyMsg, ok := msg.(legacytx.LegacyMsg); ok {
```

Comment:
Am I right in saying that the `else if` path will never be called anymore?

Assuming all modules are correctly wired by the app developer, all Msgs should be registered in the `msgServiceRouter`, so the legacy handlers will never be called.

Reply:
I believe that's true, but it may not be true for other users of the SDK. If we remove this from `BaseApp`, we should remove `AppModule.Route`. We can either do this in 0.43 or postpone till the next release. Wdyt?

Reply:
Let's maybe leave like this for v043. Probably a better time to do this is after we decide what to do with https://github.com/cosmos/cosmos-sdk/discussions/9182, which might change AppModule.